File: llm_written_v1/gemini-2.5-pro/sccs/python/algo_sequential.py
from typing import List, Set, Tuple

class Graph:
    def __init__(self, v: int):
        self.V: int = v
        self.adj: List[List[int]] = [[] for _ in range(v)]
        self.rev_adj: List[List[int]] = [[] for _ in range(v)]

    def add_edge(self, v: int, w: int) -> None:
        self.adj[v].append(w)
        self.rev_adj[w].append(v)

    def tarjan_dfs_iterative(
        self,
        start_node: int,
        disc: List[int],
        low: List[int],
        stack: List[int],
        in_stack: List[bool],
        time_ref: List[int],
        scc_list: List[List[int]],
    ):
        """Iterative version of Tarjan's DFS to avoid recursion depth errors."""
        call_stack = [(start_node, 0)]  # (node, neighbor_index)

        while call_stack:
            u, neighbor_idx = call_stack[-1]

            if neighbor_idx == 0:
                # First time visiting u
                time_ref[0] += 1
                disc[u] = low[u] = time_ref[0]
                stack.append(u)
                in_stack[u] = True

            if neighbor_idx < len(self.adj[u]):
                # Process next neighbor
                v = self.adj[u][neighbor_idx]
                call_stack[-1] = (u, neighbor_idx + 1) # Update parent state

                if disc[v] == -1:
                    call_stack.append((v, 0)) # Recurse on child
                elif in_stack[v]:
                    low[u] = min(low[u], disc[v])
            else:
                # Finished with all of u's neighbors, process u
                call_stack.pop()
                if low[u] == disc[u]:
                    scc: List[int] = []
                    while True:
                        w = stack.pop()
                        in_stack[w] = False
                        scc.append(w)
                        if w == u:
                            break
                    scc_list.append(scc)
                
                # After processing, update parent's low-link value
                if call_stack:
                    parent, _ = call_stack[-1]
                    low[parent] = min(low[parent], low[u])

    def find_sccs(self) -> List[List[int]]:
        disc = [-1] * self.V
        low = [-1] * self.V
        in_stack = [False] * self.V
        stack: List[int] = []
        scc_list: List[List[int]] = []
        time_ref = [0]

        for i in range(self.V):
            if disc[i] == -1:
                self.tarjan_dfs_iterative(i, disc, low, stack, in_stack, time_ref, scc_list)

        return scc_list

    def minimize_edges_in_scc(self, scc: List[int]) -> List[Tuple[int, int]]:
        if not scc:
            return []
        nodes: Set[int] = set(scc)
        
        forward_tree = self.build_spanning_tree(scc[0], self.adj, nodes)
        reverse_tree = self.build_spanning_tree(scc[0], self.rev_adj, nodes)
        essential_edges = list(forward_tree.union(reverse_tree))

        return essential_edges

    def build_spanning_tree(
        self, start: int, graph: List[List[int]], nodes: Set[int]
    ) -> Set[Tuple[int, int]]:
        spanning_tree: Set[Tuple[int, int]] = set()
        visited: Set[int] = set()
        stack: List[int] = [start]
        visited.add(start)

        while stack:
            node = stack.pop()
            for neighbor in graph[node]:
                if neighbor in nodes and neighbor not in visited:
                    spanning_tree.add((node, neighbor))
                    visited.add(neighbor)
                    stack.append(neighbor)
        
        return spanning_tree

    def reduce_edges(self) -> List[Tuple[int, int]]:
        SCCs = self.find_sccs()
        
        reduced_edges: List[Tuple[int, int]] = []
        for scc in SCCs:
            min_edges = self.minimize_edges_in_scc(scc)
            reduced_edges.extend(min_edges)

        # Sort for deterministic output
        reduced_edges.sort()
        return reduced_edges

File: archive/llm_written_v1/gemini-2.5-pro/sccs/cpp/algo_sequential.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
#include <unordered_set>
#include <utility>

using namespace std;

// A simple hash for pairs to be used in unordered_set
struct PairHash_Seq {
    size_t operator()(const pair<int,int>& p) const noexcept {
        return (static_cast<size_t>(p.first) << 32) ^ static_cast<size_t>(p.second);
    }
};

class GraphSequential {
private:
    int V;
    vector<vector<int>> adj, revAdj;

    void TarjanDFS(
        int u,
        vector<int>& disc,
        vector<int>& low,
        vector<int>& stack,
        vector<bool>& inStack,
        int& timer,
        vector<vector<int>>& sccList
    ) {
        disc[u] = low[u] = ++timer;
        stack.push_back(u);
        inStack[u] = true;

        for (int v : adj[u]) {
            if (disc[v] == -1) {
                TarjanDFS(v, disc, low, stack, inStack, timer, sccList);
                low[u] = min(low[u], low[v]);
            } else if (inStack[v]) {
                low[u] = min(low[u], disc[v]);
            }
        }

        if (low[u] == disc[u]) {
            vector<int> scc;
            int w;
            do {
                w = stack.back();
                stack.pop_back();
                inStack[w] = false;
                scc.push_back(w);
            } while (w != u);
            sccList.push_back(move(scc));
        }
    }

    unordered_set<pair<int,int>, PairHash_Seq>
    BuildSpanningTree(int start, const vector<vector<int>>& graph, const unordered_set<int>& nodes) {
        unordered_set<pair<int,int>, PairHash_Seq> spanningTree;
        unordered_set<int> visited;
        vector<int> st;
        st.push_back(start);
        visited.insert(start);

        while (!st.empty()) {
            int node = st.back(); st.pop_back();
            for (int nb : graph[node]) {
                if (nodes.count(nb) && !visited.count(nb)) {
                    spanningTree.insert({node, nb});
                    visited.insert(nb);
                    st.push_back(nb);
                }
            }
        }
        return spanningTree;
    }

public:
    GraphSequential(int v) : V(v), adj(v), revAdj(v) {}

    void AddEdge(int v, int w) {
        adj[v].push_back(w);
        revAdj[w].push_back(v);
    }

    vector<vector<int>> FindSCCs() {
        vector<int> disc(V, -1), low(V, -1);
        vector<bool> inStack(V, false);
        vector<int> stack;
        vector<vector<int>> sccList;
        int timer = 0;

        for (int i = 0; i < V; ++i) {
            if (disc[i] == -1) {
                TarjanDFS(i, disc, low, stack, inStack, timer, sccList);
            }
        }
        return sccList;
    }

    vector<pair<int,int>> MinimizeEdgesInSCC(const vector<int>& scc) {
        if (scc.empty()) {
            return {};
        }
        unordered_set<int> nodes(scc.begin(), scc.end());
        vector<pair<int,int>> essentialEdges;

        auto forwardTree = BuildSpanningTree(scc[0], adj, nodes);
        auto reverseTree = BuildSpanningTree(scc[0], revAdj, nodes);

        essentialEdges.reserve(forwardTree.size() + reverseTree.size());
        for (const auto& e : forwardTree) essentialEdges.push_back(e);
        for (const auto& e : reverseTree) essentialEdges.push_back(e);

        return essentialEdges;
    }

    vector<pair<int,int>> ReduceEdges() {
        auto SCCs = FindSCCs();
        vector<pair<int,int>> reducedEdges;
        for (const auto& scc : SCCs) {
            if (scc.size() > 1) {
                auto minEdges = MinimizeEdgesInSCC(scc);
                reducedEdges.insert(reducedEdges.end(), minEdges.begin(), minEdges.end());
            }
        }
        
        sort(reducedEdges.begin(), reducedEdges.end());
        return reducedEdges;
    }
};
